File: src/test/java/net/sf/rails/game/state/WalletBagTest.java
package net.sf.rails.game.state;

import static org.junit.Assert.*;

import net.sf.rails.game.state.Countable;
import net.sf.rails.game.state.Owner;
import net.sf.rails.game.state.Root;
import net.sf.rails.game.state.WalletBag;

import org.junit.Before;
import org.junit.Test;

public class WalletBagTest {

    private final static String WALLET_A_ID = "WalletA";
    private final static String WALLET_B_ID = "WalletB";

    private final static String OWNER_A_ID = "OwnerA";
    private final static String OWNER_B_ID = "OwnerB";
    private final static String ITEM_ID = "Item";
    private final static int AMOUNT = 10;
  
    private Root root;
    private WalletBag<Countable> walletA;
    private WalletBag<Countable> walletB;
    private Owner ownerA;
    private Owner ownerB;
    private Countable item;

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        ownerA = OwnerImpl.create(root, OWNER_A_ID);
        ownerB = OwnerImpl.create(root, OWNER_B_ID);
        item = CountableItemImpl.create(root, ITEM_ID);
        walletA = WalletBag.create(ownerA, WALLET_A_ID , Countable.class, item);
        walletB = WalletBag.create(ownerB, WALLET_B_ID , Countable.class, item);
        StateTestUtils.close(root);
    }

    @Test
    public void test() {
        assertEquals(0, walletA.value());
        item.move(ownerA, AMOUNT, ownerB);
        assertEquals(-AMOUNT, walletA.value());
        assertEquals(AMOUNT, walletB.value());
        
        StateTestUtils.closeAndUndo(root);
        assertEquals(0, walletA.value());
        assertEquals(0, walletB.value());
        
        StateTestUtils.redo(root);
        assertEquals(-AMOUNT, walletA.value());
        assertEquals(AMOUNT, walletB.value());
    }
    
    
}

File: dotnet/GameStateTest/WalletBagTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class WalletBagTest
    {
        private const string WALLET_A_ID = "WalletA";
        private const string WALLET_B_ID = "WalletB";

        private const string OWNER_A_ID = "OwnerA";
        private const string OWNER_B_ID = "OwnerB";
        private const string ITEM_ID = "Item";
        private const int AMOUNT = 10;

        private Root root;
        private WalletBag<ICountable> walletA;
        private WalletBag<ICountable> walletB;
        private IOwner ownerA;
        private IOwner ownerB;
        private ICountable item;

        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();
            ownerA = OwnerImpl.Create(root, OWNER_A_ID);
            ownerB = OwnerImpl.Create(root, OWNER_B_ID);
            item = CountableItemImpl.Create(root, ITEM_ID);
            walletA = WalletBag<ICountable>.Create(ownerA, WALLET_A_ID, item);
            walletB = WalletBag<ICountable>.Create(ownerB, WALLET_B_ID, item);
            StateTestUtils.Close(root);
        }

        [TestMethod]
        public void TestWalletBag()
        {
            Assert.AreEqual(0, walletA.Value());
            item.Move(ownerA, AMOUNT, ownerB);
            Assert.AreEqual(-AMOUNT, walletA.Value());
            Assert.AreEqual(AMOUNT, walletB.Value());

            StateTestUtils.CloseAndUndo(root);
            Assert.AreEqual(0, walletA.Value());
            Assert.AreEqual(0, walletB.Value());

            StateTestUtils.Redo(root);
            Assert.AreEqual(-AMOUNT, walletA.Value());
            Assert.AreEqual(AMOUNT, walletB.Value());
        }

    }
}
